docs/internals.rst

```diff
@@ -416,6 +400,84 @@ The method returns ``True`` if the permission is granted, ``False`` if denied.
 
+.. _datasette_resource_for_action:
+
+.resource_for_action(action, parent, child)
+------------------------------------------
+
+``action`` - string
+    The name of the action the resource should be associated with.
+
+``parent`` - string or ``None``
+    The parent identifier for that resource, such as the database name.
+
+``child`` - string or ``None``
+    The child identifier for that resource, such as the table or query name.
+
+Creates an instance of the appropriate ``Resource`` subclass for the provided
+action. ``Resource`` classes describe the instance, database, table and query
+objects that permissions can apply to. This helper looks up the action's
+``resource_class`` and instantiates it with the supplied identifiers, raising a
+``ValueError`` if the action name is not recognized.
+
+This is particularly useful when constructing resources based on incoming data
+such as query string parameters. For example::
+
+    resource = datasette.resource_for_action(
+        "view-table", parent="fixtures", child="facetable"
+    )
+    allowed = await datasette.allowed(
+        action="view-table", resource=resource, actor=request.actor
+    )
+
+.. _datasette_allowed_resources_sql:
+
+await .allowed_resources_sql(\*, action, actor=None, parent=None, include_is_private=False)
+-------------------------------------------------------------------------------------------
+
+Builds the SQL query that Datasette uses to determine which resources an actor
+may access for a specific action. Returns a ``(sql, params)`` tuple that can be
+executed against the internal ``catalog`` database. ``parent`` can be used to
+limit results to a specific database, and ``include_is_private`` adds a column
+indicating whether anonymous users would be denied access to that resource.
+
+Plugins that need to execute custom analysis over the raw allow/deny rules can
+use this helper to run the same query that powers the ``/-/allowed`` debugging
+interface.
+
+.. _datasette_allowed_resources:
+
+await .allowed_resources(action, actor=None, \*, parent=None, include_is_private=False)
+--------------------------------------------------------------------------------------
+
+Returns a list of ``Resource`` objects that the actor can access for the
+specified action. Each returned object is an instance of the action's
+``resource_class`` and may include a ``.private`` attribute (when
+``include_is_private=True``) to indicate that anonymous actors would be denied
+access.
+
+Example::
+
+    tables = await datasette.allowed_resources(
+        "view-table", actor=request.actor, parent="fixtures"
+    )
+    for table in tables:
+        print(table.parent, table.child)
+
+This method uses :ref:`datasette_allowed_resources_sql` under the hood and is an
+efficient way to list the databases, tables or queries visible to a user.
+
+.. _datasette_allowed_resources_with_reasons:
+
+await .allowed_resources_with_reasons(action, actor=None)
```

Comment:
Maybe this one should be called something like `*_details` and should do the is private thing AND the reasons thing?